### PR TITLE
Improved generation & bug fixes

### PR DESCRIPTION
**Pull request by BryceBarbara**
Please let me know what you think and if there are any changes I should make.

## Changes

I did a lot of refactoring (there's still so much left to do) but I mostly tried to get it at least consistently output results. I ran into a number of issues when initially trying this project so I wanted to help improve it for those who try it in the future.

### Behavior Changes

* Added support for endpoints that have no response content
* Added support for endpoints that return raw strings
* Fixed issues when operations have no `operationId` by generating some place holder one based on the path
* Fixed a crash when trying to write to an output directory that doesn't exist
* ~~Disabled exporting all the types in `api.ts`. This was to fix some ident issues it caused~~ fixed it!
* The `pick` function will no longer be emitted if it won't be used
* An operation's result type will be `unknown` if it should have a body but doesn't specify the schema

### CODE REVIEW

**Comment by Eywek** [comment]
Hello

Thank you for your contribution!
Here's my feedback:

- It seems that you haven't run tslint, can you fix linting issues? You can run the following command to view issues:
	```
	yarn tslint --project .
	```
- What is the issue with exporting types? In my company, where we're using this library, we use exported types in some use cases, removing the export is not something we want

**Comment by BryceBarbara** [comment]
No worries! I'll enable the exporting of types again, I just ran into some odd behavior with `ts-morph` incorrectly indenting everything after it. After like 30 minutes of fighting with it, I just disabled it until I wanted to dive in again.

And yeah, I'll run linting and get it all fixed up! Sorry, I'm used to eslint so I just assumed it wasn't setup yet rather than being a different system.

**Comment by BryceBarbara** [comment]
I created [an issue](https://github.com/dsherret/ts-morph/issues/1169) on the `ts-morph` about the indent issue caused by the type exports. Now we wait lol

**Comment by BryceBarbara** [comment]
@Eywek I refactored a bit more (splitting things into their own files) and fixed the type exporting with a workaround. Can you take a peek when you get time and let me know what you think?

**Comment by Eywek** [comment]
Thank you, I'll take time to review it this week-end

File: src/api.ts [hunk 1]
```diff
@@ -1,206 +1,276 @@
...
   file.addImportDeclaration({
     defaultImport: 'axios',
     namedImports: ['AxiosInstance', 'AxiosRequestConfig'],
     moduleSpecifier: 'axios'
   })
-  file.addImportDeclaration({
-    namespaceImport: 'Types',
-    moduleSpecifier: './definitions'
-  })
+  if (referencedTypes.length) {
```
Comment (Eywek) [inline src/api.ts:173]: You also need to import `WithoutWriteonly<>` and `WithoutReadonly<>` types if we need them (note: you also need to export them).

File: src/api.ts [hunk 2]
```diff
@@ -1,206 +1,276 @@
...
+  options: GenerateApiOptions
+}
+
+export interface GenerateApiOptions {
+  removeTagFromOperationId: boolean
+  addReadonlyWriteonlyModifiers: boolean
+}
+
+function getAllReferencedTypes (analyzedPaths: AnalyzedPathItemObject[]) {
+  const referencedTypeNames: Record<string, boolean> = {}
```
Comment (Eywek) [inline src/api.ts]: IMO you could do a `new Set()` instead of an object since you only need keys

File: src/api.ts [hunk 3]
```diff
@@ -1,206 +1,276 @@
...
+  spec: OpenAPIV3.Document
+  options: GenerateApiOptions
+}
+
+export interface GenerateApiOptions {
+  removeTagFromOperationId: boolean
+  addReadonlyWriteonlyModifiers: boolean
+}
+
+function getAllReferencedTypes (analyzedPaths: AnalyzedPathItemObject[]) {
```
Comment (Eywek) [inline src/api.ts]: It seems that you miss some types (i.e. the `Company` type in example), you can see errors with the example if you run `ts-node -T cmd/oatyp.ts -s example/openapi.yaml -d example -r true` and check `example/api.ts` file

**Review by Eywek — CHANGES_REQUESTED** [review]
Thank you again for those changes!

Everything seems fine for me except my few review comments (obviously since I've been lazy and there is no tests, maybe we introduced some bugs that I haven't caught but it's fine for me currently)

**Comment by BryceBarbara** [comment]
@Eywek All addressed! I also switched the package used for YAML parsing since the previous one had issues with escaped quotes and a few other things. Let me know if you have any additional changes you'd like to see!

**Comment by Eywek** [comment]
Thank you!
Can you resolve the conflict I just made?
I will review this in the week and hopefully merge it!

**Comment by BryceBarbara** [comment]
@Eywek Done! :)

**Comment by BryceBarbara** [comment]
Once this gets merged, I'll work on transitioning to using eslint rather than the deprecated tslint

File: src/code-gen.ts [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        return enumNodes.length === 1
+          ? enumNodes[0]
+          : ts.factory.createUnionTypeNode(enumNodes)
+      }
+      return CodeGen.nullableNodeType(
+        ts.factory.createKeywordTypeNode(ts.SyntaxKind.StringKeyword),
+        !!schema.nullable
+      )
+    }
+
```
Comment (Eywek) [inline src/code-gen.ts:200]: I think here we miss the https://github.com/Eywek/oatyp/commit/bf0a7dbc11b6cc33908afece33190eec9c1a22a2 fix

Comment (BryceBarbara) [reply src/code-gen.ts:200]: Added it! Nice catch! I must've missed it somehow

**Review by Eywek — APPROVED** [review]
Everything seems fine, I hope we haven't created any bugs. Thank you!

**Comment by BryceBarbara** [comment]
@Eywek Woo! 🎉

If we did introduce bugs, they shouldn't be too hard to fix :)

**Pull request merged**